[big merge][data] Combine index-based masking with student-teacher training, and pass ModelBatch class

shmh40 opened the pull request:
## Description

Combining index-based masking with student-teacher data generation, and passing the model batch class.


## Issue Number

Closes #1270 
Closes #1108 
Closes #1242 

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [ ] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

clessig committed:
NOT WORKING: initial draft for index-based masking. Implemented for random and healpix masking. Open issues with _coords_local, centroids and probably other things.

clessig committed:
NOT WORKING: Finished src, target still to be done.

clessig committed:
Masking target is working in principle but errors when feeding data to the model.

clessig committed:
Working version for ERA5, NPP-ATMS. Problems with SYNOP with empty cell handling

clessig committed:
Minor cleanup

clessig committed:
Fixed linting

clessig committed:
Fixed remaining problems that occured for NPP-ATMS and SYNOP.
TODO:
- Forecast still needs to be adapted
- Some more cleanup of variable naming, return values etc

clessig committed:
Enabled support for forecast. Cleaned up some bits and pieces.

clessig committed:
Removing centroids options for embedding that was unused and should not be used.

clessig committed:
Removed unused parameters

clessig committed:
Inversion of target output ordering to match input one in forcast mode. Unclear how to deal with it with MTM

clessig committed:
Changes to  prepare_logging to apply index inversion

shmh40 committed:
added file with ModelBatch and SampleMetadata dataclasses

clessig committed:
Updating config to working version

shmh40 committed:
update ViewMetadata spec

shmh40 committed:
draft changes to allow global local view generation in masker and tokenizer_masking. generate the mask, otherwise using batchify_source and batchify_target as before, with the capacity to remember what mask we have now when it comes to generating the targets. Update to inputs_metadata structure but not put in to practice

shmh40 committed:
draft of training_config in default_config

shmh40 committed:
change view_metadata to dict in ModelInput

clessig committed:
NOT WORKING: updating class to handle multiple input steps and improving overall structure

clessig committed:
Merge branch 'shmh40/dev/1270-idx-global-local' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Added basic support for multi-step sources.

clessig committed:
Partially enabled correct handling of multiple input steps.

clessig committed:
Added mode and refactored get_sample_data into separate function.

clessig committed:
Comments

clessig committed:
Renaming

shmh40 committed:
updated default config training_config to allow student-teacher

shmh40 committed:
added stream id to era5 config

shmh40 committed:
slight restructure of ViewMetadata

shmh40 committed:
basic if statement to yield the student and teacher views

shmh40 committed:
merge changes

shmh40 committed:
correct imports with new batch.py

shmh40 committed:
created function for _get_student_teacher_sample_data which returns the streams_data of the teacher and multiple streams_datas for the student views.

clessig committed:
Not working draft for restructuring

clessig committed:
Merge branch 'shmh40/dev/1270-idx-global-local' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Changes for better student teacher structure

clessig committed:
More refactoring

clessig committed:
More refactoring and cleanup

clessig committed:
More refactoring. Code working again.

clessig committed:
Cleaned up parametrization

clessig committed:
Changes necessary for spoofing flag per IOReaderData

clessig committed:
Changes to have spoofing on a per data reader sample

clessig committed:
Moved _get_student_teacher_masks() so that masks are generated for all streams first.

clessig committed:
Renaming and minor clean up.

clessig committed:
Added basic support for use of ModelBatch class to define rough structure and interface.

clessig committed:
linting

clessig committed:
Linting

clessig committed:
linting

clessig committed:
Linting problems but removed unused ViewMetaData dependence

clessig committed:
Added required reflexivity between source and target samples to Batch

clessig committed:
Added todo

shmh40 committed:
fix typo in ModelBatch

shmh40 committed:
collect num_source_samples and num_target_samples, add loop over teacher masks hence allowing multiple teacher views, and add source_target_idx to keep track of which student belongs to which teacher

shmh40 committed:
add teacher num_views parameter to config

clessig committed:
Re-enabling inversion of targert ordering.

shmh40 committed:
tidy up, remove unused build_stream_views in tokenizer_masking

shmh40 committed:
multiple idxs for each teacher, need to confirm for not student case, and updated ModelBatch for this

tjhunter commented on src/weathergen/datasets/batch.py:
```diff
@@ -0,0 +1,180 @@
+    masking_strategy: str
+
+    # parameters for masking strategy
+    masking_params: dict
```
same: `Config`

tjhunter commented on src/weathergen/datasets/masking.py:
```diff
@@ -139,6 +135,113 @@ def _select_strategy(self):
 
+    def mask_source_idxs(
+        self,
+        stream_info,
```
note you are not using most of the arguments

tjhunter commented on src/weathergen/datasets/masking.py:
```diff
@@ -139,6 +135,113 @@ def _select_strategy(self):
+        return (self.mask_tokens, self.mask_channels)
+
+    def mask_targets_idxs(
+        self,
```
likewise

tjhunter commented on src/weathergen/datasets/multi_stream_data_sampler.py:
```diff
@@ -317,16 +297,388 @@ def reset(self):
+    def _get_source_target_masks(self, idx: int, forecast_dt: int):
+        """
+        Return one batch of data
+        Build a StreamData object for a single view (teacher or student).
```
inaccurate

tjhunter commented on src/weathergen/datasets/tokenizer_masking.py:
```diff
@@ -7,24 +7,35 @@
+    """
+    Convert data, coords, and geoinfos to torch tensor
+    """
+    if type(rdata.coords) is not torch.Tensor:
```
don't update in place, the signature of the function assumes your function is pure. use `dataclasses.replace`. You don't even need to check if it is a tensor. In that case, it would be a no-op:

```py
return replace(rdata, coords=torch.tensor(rdata.coords), ...)
```

tjhunter commented on src/weathergen/datasets/tokenizer_masking.py:
```diff
@@ -34,187 +45,219 @@ def __init__(self, healpix_level: int, masker: Masker):
 
-        target_tokens = self.masker.mask_target(
-            target_tokens_cells, rdata.coords, rdata.geoinfos, rdata.data
+        data, datetimes, coords, coords_local, coords_per_cell = tokenize_apply_mask_target(
```
good place for a dataclass

tjhunter commented on src/weathergen/datasets/utils.py:
```diff
@@ -267,356 +266,7 @@ def add_local_vert_coords_ctrs2(verts_local, tcs_lens, a, zi, geoinfo_offset):
     return a
 
 
-####################################################################################################
```
thank you for the cleanup

tjhunter commented on src/weathergen/datasets/view_builder.py:
```diff
@@ -0,0 +1,97 @@
+    teacher_cfg: dict,
+    student_cfg: dict,
+    relationship: str = "subset",
+) -> Tuple[np.ndarray, List[np.ndarray], List[ViewMetadata]]:
```
np.typing.NDArray

tjhunter reviewed (commented):
A first batch of comments.

In general:
- put more types, especially for dicts. It helps following the flow of data. We have a complex processing logic
- I would add more logging to show the relationship between the different view ids and student/teach ids
- I would use less indexes and boolean masks, and instead use the view / stream ids. It simplifies teh logic and does not at this stage introduce perf issues

shmh40 committed:
add max_num_targets to era5

shmh40 committed:
add max_num_samples functionality to tokenizer_masking and pass through in multi_stream_data_sampler. coords_per_cell is a bit nasty

shmh40 committed:
move build_views_for_stream into masker

shmh40 commented on config/default_config.yml:
```diff
@@ -113,8 +108,41 @@ masking_strategy_config: {"strategies": ["random", "healpix", "channel"],
+# TODO: adapt so that the masking or forecast config entry also sits here
+training_config:
+  # when this is "masking", we are basically only using the model_input subconfig
+  training_mode: "student_teacher"  # "masking", "student_teacher", "forecast"
```
I understand this is perhaps not perfectly delineated, but this mode will define the training and loss that will be used, so training_mode seems ok to me.

shmh40 commented on config/default_config.yml:
```diff
@@ -113,8 +108,41 @@ masking_strategy_config: {"strategies": ["random", "healpix", "channel"],
+  model_input:
+    masking_strategy: "healpix" # "random", "healpix". Masking strategy to use for model input for masking, and local (student) views when doing student-teacher
+    rate: 0.5  # Masking rate to use for model input
+    num_views: 4  # if student-teacher, the number of local (student) views to generate
```
Agreed. I need to think about how best to do this with @sophie-xhonneux , though this suggestion seems reasonable.

shmh40 commented on src/weathergen/datasets/batch.py:
```diff
@@ -0,0 +1,180 @@
+    Attributes:
+        view_id: Unique identifier (e.g., "teacher_global", "student_local_0")
+        keep_mask: Boolean array [num_healpix_cells] at data level indicating kept cells
+        strategy: Name of selection strategy ("random", "healpix_level_2", etc.)
```
Agreed, this is going to be deprecated

shmh40 committed:
tidy up, remove unused arguments, types

shmh40 committed:
fix masking for NPP-ATMS by correctly selecting final timestep mask and aligning between source and target. working for num_input_steps = 1, broken for > 1, compute_offsets_scatter_embed not working

shmh40 committed:
merge with develop, include trainer idx_inv_rt, merged default_config, rm tokenizer_forecast

shmh40 committed:
updated configs so code runs. Note default config to be overhauled still

shmh40 committed:
very hacky first pass of full masking_strategy_config for the student and teacher views. Much to fix up

shmh40 committed:
instructions for sophie

shmh40 committed:
add SampleMetaData integration and functionality, and update masker to use SampleMetadata. Pass through source_cell_lens and target_coords_idx to student_teacher_batch in iter, and hence pass through to trainer. source_cell_lens and target_coords_idx are now part of Sample, which is itself the components of ModelBatch. To tidy

shmh40 committed:
remove prints, pdb

shmh40 committed:
add mask to SampleMetaData and add forecast_dt to Sample so it is accessible. Can specify the loss in the default config with student-teacher views

shmh40 committed:
add diffusion forecast option for the data sampling, and with noise_level_rn in the metadata. The Trainer needs to be copied from Sophies branch, currently we only get so far

clessig committed:
Merge branch 'shmh40/dev/1270-idx-global-local' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Linting

clessig committed:
Simplified and clarified handling of default target_aux_calcualtor

clessig committed:
Linting

clessig committed:
Linting

clessig committed:
Linting

clessig committed:
Linting

clessig committed:
Linting

clessig committed:
Restoring masking as training_mode in default_config

clessig committed:
More linting

clessig committed:
Removed duplicate lines due to mergeing

clessig committed:
Restored masking as training mode. Not working due to NaN in prediction

clessig committed:
Fixed problem in engines introduced in recent commits merging develop. This fixes masking training

shmh40 committed:
remove unused mask generation in diffusion_forecast

shmh40 committed:
restore masking_strategy to random

Had placeholder for testing, now back to "random" for masking strategy in the base level of default_config

shmh40 committed:
restore loader_num_workers to 8

shmh40 committed:
fix indentation of else: assert False in _get_sample msds

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

tjhunter committed:
merge with dev

tjhunter committed:
linter warnings

tjhunter committed:
commenting tests

clessig committed:
Restructured code so that mask generation and application is cleanly separated

clessig committed:
Commit

clessig committed:
Update

clessig committed:
Merge branch 'shmh40/dev/1270-idx-global-local' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Fixed uv.lock

clessig committed:
Fix for integration test

clessig committed:
Re-enabled multi-source training

Jubeku committed:
1390 - Adapt forward pass of new batch object (#1391)

* Add to device to ModelBatch, etc & adapt model

TODO adapt validate and inference
TODO test forecasting and multiple stream because predict changed
substantially

* Rename view to sample and fix validate

* Revert predict function and fix inference

* Fix invalid access with mask

* Linting

* Fixed handling of target_idxs and other minor issues

---------

Co-authored-by: sophiex <24638638+sophie-xhonneux@users.noreply.github.com>
Co-authored-by: Christian Lessig <christian.lessig@ecmwf.int>

clessig committed:
Completed migration to new batch class by removing reference to old list of lists

clessig committed:
Fixed missing non_blocking=True in to_device()

clessig committed:
Removed old comments

clessig committed:
Fixed problem with non_blocking=True

clessig committed:
Cleaned up comments and return values a bit

clessig committed:
Changed args to embedding

clessig committed:
Changed core functions to take sample as arg

clessig committed:
Changed that model takes sample as input

clessig committed:
Fixes for diffusion

clessig committed:
Switched to lists of model / target stratgies

clessig committed:
Updated config

clessig committed:
Changed to per masking strategy loss terms

clessig committed:
Removed old masking options. Still needs to be fully cleaned up

clessig committed:
More robust handling of empty streams

clessig committed:
Fixed incorrect handling of empty target_coords_idx

clessig committed:
Fixed problem when number of model and target samples is different

clessig committed:
Example for config with non-trivial model and target inputs

clessig committed:
Fixed bug in total sample counting

clessig committed:
Re-enabled missing healpix level

clessig committed:
Fixed incorrect handling of masking and student_teacher modes. Follow up fixes required to handle partially filler source/target streams (because source has no target values, eg).

kctezcan committed:
An encoder formed by embedding + local assimilation + global assimilation (#1397)

* initial changes

* more changes

* removed extra print parameters statement

* changed names for backward checkpoint loading

* added encoder. to module names in sharding

* adding encoder. to embed_engine

* added back the conditions for param printong

* lint

* forecast config

* switch back to MTM config

* lint

clessig committed:
Formatting

clessig committed:
Fix source-target matching problem.

clessig committed:
Enabled multiple input steps. Fixed various robustness that arose through this.

This commit also changes the number of forecast steps that are taken. The old loop was at least one step too far. Unclear why the problem occurred now.

clessig committed:
Linting

clessig committed:
Missing update to validation()

clessig committed:
Improved robustness through sanity checking of arguments

clessig committed:
Improved handling of corner cases

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
- Fixed incorrect call to get_forecast_steps() in validation
- Fixed interface of target_aux_calculator

clessig committed:
More fixed to validation

clessig committed:
Adding stream_id

clessig committed:
Cleaned up ModelOutput class to have proper access functions and a better structure

clessig committed:
Switched to use dict to internally represent streams_datasets

clessig committed:
Improving robustness of interface of ModelOutput class

clessig committed:
Re-enabling model output

clessig committed:
Ruff

clessig committed:
Minor clean-ups and additional comments

clessig committed:
Minor cleanups

clessig committed:
Cleaned up handling of masks and masking metadata

clessig committed:
Current working version of default_config

clessig committed:
Fixed problem with branches with old code and incomplete cleanup

clessig committed:
Updated to test convergence of integration test.

clessig committed:
Updated settings

clessig committed:
Clessig/ypd/dev/1353 add tokens latent state finalization (#1452)

* Add LatentState

* Add class and register tokens for LatentState, adjust everything accordingly

* Add option in config file + minor changes

* Add pos.emb. for register tokens + remove class tokens + minor fixes

* Minor fix

* Changed empty to zeros pe_register

* Ruffed

* Clean-up and fixed positional encoding

* Fixing things that got lost during last merge

---------

Co-authored-by: Yura Perugachi Diaz <yperugac@santis-ln001.cscs.ch>
Co-authored-by: Yura Perugachi Diaz <yperugac@santis-ln002.cscs.ch>

clessig committed:
Ruffed

clessig committed:
Adding sanity check for register tokens

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Improved strucutre of LatentState class.

clessig committed:
Improved structure of LatentState

clessig committed:
Fixed problem wiht num_samples > 1

clessig committed:
Improved representation of batch and batch data and more getter functions

clessig committed:
Re-enabled batch_size/num_samples>1. Still some minor problems and cleanup needed but overall program flow working

clessig committed:
Cleanup

clessig committed:
Fixed bug in source-target correspondence with num_samples>1

clessig committed:
Removing incorrect loss scaling

simone99n commented on integration_tests/small_multi_stream.yaml:
```diff
@@ -4,12 +4,11 @@ model_path: "./models"
 num_mini_epochs: 1
-samples_per_mini_epoch: 16
+samples_per_mini_epoch: 128
 samples_per_validation: 4
```
`samples_per_validation` must be divisible by  `loader_num_workers`, otherwise it will fail

clessig committed:
Cleaned up predict() in model

clessig committed:
Fixed commenting issues

clessig committed:
Fixed problem with freezing of modules for q_cells. Fixed problem when runing in FSDP

clessig committed:
Fixed problem with printing of trainable weights

clessig committed:
Fixed switch for printing of trainable weights

Jubeku committed:
1316 Update metrics logging (#1412)

* Add to device to ModelBatch, etc & adapt model

TODO adapt validate and inference
TODO test forecasting and multiple stream because predict changed
substantially

* Rename view to sample and fix validate

* Revert predict function and fix inference

* Fix invalid access with mask

* Linting

* Fixed handling of target_idxs and other minor issues

* Remove duplicate to_device

* move loss history into loss calculator

* handle loss_avg and unflatten loss dict

* fixing train_logger

* update validate logging, failing - need to merge data branch

* rm additional log files and log_vals variable, and collapse to single add_logs fct for train and val

* rm comment

* fix validation

* move prepare losses fct to train_logger script, fix terminal logging for val

* fix ctr_loss_fcts normalization; calculate per stream, per lfct average across channels and fsteps for logging

* Fixed linting

* fix bug in emptying history after logging

---------

Co-authored-by: sophiex <24638638+sophie-xhonneux@users.noreply.github.com>
Co-authored-by: Christian Lessig <christian.lessig@ecmwf.int>

Jubeku committed:
set target_channels dependent on val or train (#1471)

shmh40 committed:
fix targetauxoutput and inference to accept target times and coords. Still lists, to be changed to dicts

shmh40 committed:
updated targetauxbase so we access target, times and coords with a dict and corresponding changes

clessig committed:
Patch for float in loss dict (shouldn't happen to begin with)

clessig committed:
Merge branch 'shmh40/dev/1270-idx-global-local' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Fixed handling of per-batch precomputation. Functional for num_sample/batch_size=1 but convergence seems brokens for num_samples>1.

clessig committed:
Reordering

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Linting

clessig committed:
More linting

clessig committed:
More linting

clessig committed:
Removed old, commented code

sophie-xhonneux committed:
Push current progress for inspection (#1478)

* Push current progress for inspection

* For Seb

* Delete old code

* Rename according to config

* Fix config

* Push what I have

* Successfully build data

* Fix bugs and lint

* Forgot to revert dataloader workers

* Address PR review comments

- rename student-to-teacher
- extract metadata extraction into a function

* prepare branch for ssl merge, ready for data merge

* Removed mock-up

---------

Co-authored-by: Sebastian Hickman <seb.hickman@gmail.com>
Co-authored-by: Christian Lessig <christian.lessig@ecmwf.int>

clessig committed:
Merge branch 'shmh40/dev/1270-idx-global-local' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Cleaned up
